modules/annex/annex_archive.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package annex
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"code.gitea.io/gitea/modules/git"
+)
+
+func repoArchiveTargz(repo *git.Repository, target io.Writer, prefix, commitID string) error {
+	// create a plain git tar archive
+	var tarB bytes.Buffer
+	err := repo.CreateArchive(repo.Ctx, git.TAR, &tarB, prefix != "", commitID)
+	if err != nil {
+		return err
+	}
+
+	gitFilesR := tar.NewReader(&tarB)
+
+	gzipW := gzip.NewWriter(target)
+	defer gzipW.Close()
+	tarW := tar.NewWriter(gzipW)
+	defer tarW.Close()
+
+	tree, err := repo.GetTree(commitID)
+	if err != nil {
+		return err
+	}
+
+	for {
+		oldHeader, err := gitFilesR.Next()
+		// TODO: handle non-local names in tar archives?
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+
+		// default to copying the current file from the archive created by repo.CreateArchive
+		header := oldHeader
+		dataR := io.Reader(gitFilesR)
```

Comment:
I would expect `io.Reader` to conflict with the `.Next()` above. How does `io.Copy()` below know to stop at the end of the current file?

Obviously you got it working. Don't feel pressured to explain, I can do my own reading up in the go docs to figure out why this works.

Reply:
So the `tar.Reader` type is a bit special in that it is very stateful: you call `.Next` to advance to the next file in the archive and get back the header belonging to that file. Afterwards you can read from the `tar.Reader` using the standard `io.Reader` interface and get back the data for that file, until you encounter EOF, i.e. the end of that file.

From the docs:
> Read reads from the current file in the tar archive. It returns (0, io.EOF) when it reaches the end of that file, until Next is called to advance to the next file.

Then you can call `.Next` again and continue with the next file, until you encounter EOF from `.Next`, which marks the end of the archive.

`io.Copy` just copies everything until it encounters EOF:
> Copy copies from src to dst until either EOF is reached on src or an error occurs.

The reason I cast to `io.Reader` here is that this way I can re-assign the reader for the annexed file to this variable later - if the file is annexed - and only have to code the `io.Copy` once.